README.rst
Welcome to the python ebaysdk
=============================

This SDK is a programmatic interface into the eBay APIs. It simplifies development and cuts development time by standardizing calls, response processing, error handling, and debugging across the Finding, Shopping, Merchandising & Trading APIs. 

Quick Example::

    import datetime
    from ebaysdk.exception import ConnectionError
    from ebaysdk.finding import Connection

    try:
        api = Connection(appid='YOUR_APPID_HERE', config_file=None)
        response = api.execute('findItemsAdvanced', {'keywords': 'legos'})        

        assert(response.reply.ack == 'Success')  
        assert(type(response.reply.timestamp) == datetime.datetime)
        assert(type(response.reply.searchResult.item) == list)
  
        item = response.reply.searchResult.item[0]
        assert(type(item.listingInfo.endTime) == datetime.datetime)
        assert(type(response.dict()) == dict)

    except ConnectionError as e:
        print(e)
        print(e.response.dict())


Migrating from v1 to v2
-----------------------

Test for GPG

For a complete guide on migrating from ebaysdk v1 to v2 and see an overview of the additional features in v2 please read the `v1 to v2 guide`_


Getting Started
---------------

1) SDK Classes

* `Trading API Class`_ - secure, authenticated access to private eBay data.
* `Finding API Class`_ - access eBay's next generation search capabilities.
* `Shopping API Class`_ - performance-optimized, lightweight APIs for accessing public eBay data.
* `Merchandising API Class`_ - find items and products on eBay that provide good value or are otherwise popular with eBay buyers.
* `HTTP Class`_ - generic back-end class the enbles and standardized way to make API calls.
* `Parallel Class`_ - SDK support for concurrent API calls.

2) SDK Configuration

* Using the SDK without YAML configuration
  
   ebaysdk.finding.Connection(appid='...', config_file=None)

* `YAML Configuration`_ 
* `Understanding eBay Credentials`_

3) Sample code can be found in the `samples directory`_.

4) Understanding the `Request Dictionary`_.

Support
-------

For developer support regarding the SDK code base please use this project's `Github issue tracking`_.

For developer support regarding the eBay APIs please use the `eBay Developer Forums`_.

Install
-------

Installation instructions for *nix and windows can be found in the `INSTALL file`_.

License
-------

`COMMON DEVELOPMENT AND DISTRIBUTION LICENSE`_ Version 1.0 (CDDL-1.0)


.. _INSTALL file: https://github.com/timotheus/ebaysdk-python/blob/master/INSTALL
.. _COMMON DEVELOPMENT AND DISTRIBUTION LICENSE: http://opensource.org/licenses/CDDL-1.0
.. _Understanding eBay Credentials: https://github.com/timotheus/ebaysdk-python/wiki/eBay-Credentials
.. _eBay Developer Site: http://developer.ebay.com/
.. _YAML Configuration: https://github.com/timotheus/ebaysdk-python/wiki/YAML-Configuration
.. _Trading API Class: https://github.com/timotheus/ebaysdk-python/wiki/Trading-API-Class
.. _Finding API Class: https://github.com/timotheus/ebaysdk-python/wiki/Finding-API-Class
.. _Shopping API Class: https://github.com/timotheus/ebaysdk-python/wiki/Shopping-API-Class
.. _Merchandising API Class: https://github.com/timotheus/ebaysdk-python/wiki/Merchandising-API-Class
.. _HTTP Class: https://github.com/timotheus/ebaysdk-python/wiki/HTTP-Class
.. _Parallel Class: https://github.com/timotheus/ebaysdk-python/wiki/Parallel-Class
.. _eBay Developer Forums: https://forums.developer.ebay.com
.. _Github issue tracking: https://github.com/timotheus/ebaysdk-python/issues
.. _v1 to v2 guide: https://github.com/timotheus/ebaysdk-python/wiki/Migrating-from-v1-to-v2 
.. _samples directory: https://github.com/timotheus/ebaysdk-python/tree/master/samples
.. _Request Dictionary: https://github.com/timotheus/ebaysdk-python/wiki/Request-Dictionary


Changes
Changes for ebaysdk

2.2.0 Mon Apr 20 15:23:53 PDT 2020
- Forced HTTPS for finding, shopping, and trading calls
- fixed sample code so that it uses the appropriate domain
- added py38 to tox testing

2.1.6
- Update Copyright

2.1.5 Oct 13, 2017
- Bug fix release

2.1.4 Tue Aug 23 11:21:37 PDT 2016
- Merge PR from listingmirror for site-id fix
- Added a few performance tests

2.1.3
- Merged PR for XML escaping option (https://github.com/timotheus/ebaysdk-python/pull/139)
- fix unicode issue between py2 and py3
- add a unicode example in the samples dir
- fix for issue #127 (false-ish check)
- merge in py3 datatype fix from MartiONE

2.1.2 Thu Jun 25 14:54:12 PDT 2015 
- merge in rowbare PR that support file upload
- refactor order poller into a Class and implement a base Storage class
  for any domain-specific actions (e.g. inserting into a DB)
- merged in backend support for inventory management (from bimusiek)
- add backend for Business Policy API (from nickspring)
- set ssl verify to True in connection class

2.1.1 Tue Apr 14 20:16:47 PDT 2015
- merge in a pull request to fix Unicode issues
- remove PyYAML dependency
- add poller package for pulling orders

2.1.0 Mon Jul  7 10:03:36 PDT 2014
- modify install instructions
- remove grequests install dep from setup.py

2.0.0
- WARNING!! This release breaks some backward compatibility
  Read https://github.com/timotheus/ebaysdk-python/wiki/Migrating-from-v1-to-v2
- imports: Modified package structure
- execute(): Modified return value
- Switched to lxml.etree
- Added a new response class
  response.reply, response.dom(), response.dict(), response.json()
- Added ebaysdk exception classes
- Modified utils.py (dict2xml, xml2dict)

1.0.3
- rework Unicode fix
- fix logging handler

1.0.2 Sat Feb  8 20:44:00 PST 2014
- fix unicode issue is dict to xml routine
- add a Unicode test
- fix http backend

1.0.0
- Major refactor
  + changes are backward compatible unless you're subclassing any of the
    SDK classes
  + moved from PyCurl to Python Requests
  + code organization
  + standard python logging
  + add exceptions

0.1.11
- Add affiliate headers to the Shopping API backend
- https://github.com/timotheus/ebaysdk-python/issues/40
  fix datatype issue with error codes from the shopping API
  
- added py3 support thanks to Nikolay Derkach
- store response error codes in an array when processing response.
  The response error codes are accessible by the api.response_codes()
  accessor.

  if api.error():
      if 37 in api.response_codes():
          print "Invalid data"

0.1.10
- added error checking for html() class
- all class return response_data regardless of the HTTP status code
- added Merchandising API class
- update SOA class default value
- add more trading API samples

0.1.9 Wed Apr 24 14:38:40 PDT 2013
- update error handling
- add VerifyAddItem sample call
- update license in setup.py

0.1.8 Tue Apr 23 15:33:02 PDT 2013
- push dist to pypi
- fix deprecation warning in utils.py
- pep8 cleanup
- handle utf-8 when parsing with BeautifulStoneSoup
- add retry to standard & parallel calls
- a bug fix for SOA class
- add documentation and sample scripts
- clean up YAML file
- YAML values are now overridden by values defined
  when building the object
- remove silent dependency on simplejson, at least for Python >= 2.6
- fix bug in html class that allows for POST 
- refactor and style cleanup
- added parallel support using the parallel class
- created new prepare, process, and _process_http_request methods

0.1.7
- update tests
- modify response_obj() to return response_dict()

0.0.6
- support older version of Element tree
- modify dict2xml to handle nodes with array content

for example,
{ 
    'keywords': 'shoes', 
    'paginationInput': { 'pageNumber': 1 }, 
    'itemFilter': [ 
        { 'name': 'MinBids', 'value': 10 }, 
        { 'name': 'MaxBids', 'value': 14 }
    ],
}
<?xml version='1.0' encoding='utf-8'?>
<findItemsAdvancedRequest xmlns="http://www.ebay.com/marketplace/search/v1/services">
<itemFilter><name>MinBids</name><value>10</value></itemFilter>
<itemFilter><name>MaxBids</name><value>14</value></itemFilter>
<keywords>shoes</keywords>
<paginationInput><pageNumber>1</pageNumber></paginationInput>
</findItemsAdvancedRequest>

0.0.5
- added response_obj() which turns a dict into a Struct
  e.g. { a : { 'name' : 'tim' } } can be accessed like a.name
- add support to dict args in the execute call. execute will now check the 
  datatype and convert to XML if necessary
- clean doctests
    
0.0.4
- moved from httplib to pycurl
  * pycurl was chosen because of its https/proxy support and
    that it's thread-safe.

0.0.3 Wed May 11 11:03:58 PDT 2011
- fix escape username/password
- fix error handling
- cleanup doctests

0.0.2 Fri Aug 20 17:25:25 PDT 2010
- fix download link 

0.0.1 Initial Release



samples/trading.py
# -*- coding: utf-8 -*-
'''
Copyright 2012-2019 eBay Inc.
Authored by: Tim Keefer
Licensed under CDDL 1.0
'''

import os
import sys
import datetime
from optparse import OptionParser

sys.path.insert(0, '%s/../' % os.path.dirname(__file__))

from common import dump

import ebaysdk
from ebaysdk.utils import getNodeText
from ebaysdk.exception import ConnectionError
from ebaysdk.trading import Connection as Trading


def init_options():
    usage = "usage: %prog [options]"
    parser = OptionParser(usage=usage)

    parser.add_option("-d", "--debug",
                      action="store_true", dest="debug", default=False,
                      help="Enabled debugging [default: %default]")
    parser.add_option("-y", "--yaml",
                      dest="yaml", default='ebay.yaml',
                      help="Specifies the name of the YAML defaults file. [default: %default]")
    parser.add_option("-a", "--appid",
                      dest="appid", default=None,
                      help="Specifies the eBay application id to use.")
    parser.add_option("-p", "--devid",
                      dest="devid", default=None,
                      help="Specifies the eBay developer id to use.")
    parser.add_option("-c", "--certid",
                      dest="certid", default=None,
                      help="Specifies the eBay cert id to use.")
    parser.add_option("-n", "--domain",
                      dest="domain", default='api.ebay.com',
                      help="Specifies the eBay domain to use (e.g. api.sandbox.ebay.com).")

    (opts, args) = parser.parse_args()
    return opts, args


def run(opts):

    try:
        api = Trading(debug=opts.debug, config_file=opts.yaml, appid=opts.appid, domain=opts.domain,
                      certid=opts.certid, devid=opts.devid)

        api.execute('GetCharities', {'CharityID': 3897})
        dump(api)
        print(api.response.reply.Charity.Name)

    except ConnectionError as e:
        print(e)
        print(e.response.dict())


def feedback(opts):
    try:
        api = Trading(debug=opts.debug, config_file=opts.yaml, appid=opts.appid, domain=opts.domain,
                      certid=opts.certid, devid=opts.devid, warnings=False)

        api.execute('GetFeedback', {'UserID': 'tim0th3us'})
        dump(api)

        if int(api.response.reply.FeedbackScore) > 50:
            print("Doing good!")
        else:
            print("Sell more, buy more..")

    except ConnectionError as e:
        print(e)
        print(e.response.dict())


def getTokenStatus(opts):

    try:
        api = Trading(debug=opts.debug, config_file=opts.yaml, appid=opts.appid, domain=opts.domain,
                      certid=opts.certid, devid=opts.devid, warnings=False)

        api.execute('GetTokenStatus')
        dump(api)

    except ConnectionError as e:
        print(e)
        print(e.response.dict())


def verifyAddItem(opts):
    """http://www.utilities-online.info/xmltojson/#.UXli2it4avc
    """

    try:
        api = Trading(debug=opts.debug, config_file=opts.yaml, appid=opts.appid, domain=opts.domain,
                      certid=opts.certid, devid=opts.devid, warnings=False)

        myitem = {
            "Item": {
                "Title": "Harry Potter and the Philosopher's Stone",
                "Description": "This is the first book in the Harry Potter series. In excellent condition!",
                "PrimaryCategory": {"CategoryID": "377"},
                "StartPrice": "10.0",
                "BuyItNowPrice": "15.0",
                "CategoryMappingAllowed": "true",
                "Country": "US",
                "ConditionID": "3000",
                "Currency": "USD",
                "DispatchTimeMax": "3",
                "ListingDuration": "Days_7",
                "ListingType": "Chinese",
                "PaymentMethods": "PayPal",
                "PayPalEmailAddress": "tkeefdddder@gmail.com",
                "PictureDetails": {"PictureURL": "http://i1.sandbox.ebayimg.com/03/i/00/30/07/20_1.JPG?set_id=8800005007"},
                "PostalCode": "95125",
                "Quantity": "1",
                "ReturnPolicy": {
                    "ReturnsAcceptedOption": "ReturnsAccepted",
                    "RefundOption": "MoneyBack",
                    "ReturnsWithinOption": "Days_30",
                    "Description": "If you are not satisfied, return the book for refund.",
                    "ShippingCostPaidByOption": "Buyer"
                },
                "SellerProfiles": {
                    "SellerPaymentProfile": {
                        "PaymentProfileName": "PayPal:Immediate pay",
                    },
                    "SellerReturnProfile": {
                        "ReturnProfileName": "30 Day Return Policy",
                    },
                    "SellerShippingProfile": {
                        "ShippingProfileName": "USPS First Class, Priority, Priority Express Flat Rate Envelope",
                    }
                },
                "ShippingDetails": {
                    "ShippingType": "Calculated",
                    "ShippingServiceOptions": {
                        "ShippingServicePriority": "1",
                        "ShippingService": "USPSMedia"
                    },
                    "CalculatedShippingRate": {
                        "OriginatingPostalCode": "95125",
                        "PackagingHandlingCosts": "0.0",
                        "ShippingPackage": "PackageThickEnvelope",
                        "WeightMajor": "1",
                        "WeightMinor": "0"
                    }
                },
                "Site": "US"
            }
        }

        api.execute('VerifyAddItem', myitem)
        dump(api)

    except ConnectionError as e:
        print(e)
        print(e.response.dict())


def verifyAddItemErrorCodes(opts):
    """http://www.utilities-online.info/xmltojson/#.UXli2it4avc
    """

    try:
        api = Trading(debug=opts.debug, config_file=opts.yaml, appid=opts.appid, domain=opts.domain,
                      certid=opts.certid, devid=opts.devid, warnings=False)

        myitem = {
            "Item": {
                "Title": "Harry Potter and the Philosopher's Stone",
                "Description": "This is the first book in the Harry Potter series. In excellent condition!",
                "PrimaryCategory": {"CategoryID": "377aaaaaa"},
                "StartPrice": "1.0",
                "CategoryMappingAllowed": "true",
                "Country": "US",
                "ConditionID": "3000",
                "Currency": "USD",
                "DispatchTimeMax": "3",
                "ListingDuration": "Days_7",
                "ListingType": "Chinese",
                "PaymentMethods": "PayPal",
                "PayPalEmailAddress": "tkeefdddder@gmail.com",
                "PictureDetails": {"PictureURL": "http://i1.sandbox.ebayimg.com/03/i/00/30/07/20_1.JPG?set_id=8800005007"},
                "PostalCode": "95125",
                "Quantity": "1",
                "ReturnPolicy": {
                    "ReturnsAcceptedOption": "ReturnsAccepted",
                    "RefundOption": "MoneyBack",
                    "ReturnsWithinOption": "Days_30",
                    "Description": "If you are not satisfied, return the book for refund.",
                    "ShippingCostPaidByOption": "Buyer"
                },
                "ShippingDetails": {
                    "ShippingType": "Flat",
                    "ShippingServiceOptions": {
                        "ShippingServicePriority": "1",
                        "ShippingService": "USPSMedia",
                        "ShippingServiceCost": "2.50"
                    }
                },
                "Site": "US"
            }
        }

        motors_item = {
            'Item': {
                'Category': '101',
                'Title': 'My Title',
                'ItemCompatibilityList': {
                    'Compatibility': [
                        {
                            'CompatibilityNotes': 'Fits for all trims and engines.',
                            'NameValueList': [
                                {'Name': 'Year', 'Value': '2001'},
                                {'Name': 'Make', 'Value': 'Honda'},
                                {'Name': 'Model', 'Value': 'Accord'}
                            ]
                         },
                    ]
                }
            }
        }

        api.execute('VerifyAddItem', myitem)

    except ConnectionError as e:
        # traverse the DOM to look for error codes
        for node in api.response.dom().findall('ErrorCode'):
            print("error code: %s" % node.text)

        # check for invalid data - error code 37
        if 37 in api.response_codes():
            print("Invalid data in request")

        print(e)
        print(e.response.dict())


def uploadPicture(opts):

    try:
        api = Trading(debug=opts.debug, config_file=opts.yaml, appid=opts.appid, domain=opts.domain,
                      certid=opts.certid, devid=opts.devid, warnings=True)

        pictureData = {
            "WarningLevel": "High",
            "ExternalPictureURL": "http://developer.ebay.com/DevZone/XML/docs/images/hp_book_image.jpg",
            "PictureName": "WorldLeaders"
        }

        api.execute('UploadSiteHostedPictures', pictureData)
        dump(api)

    except ConnectionError as e:
        print(e)
        print(e.response.dict())


def uploadPictureFromFilesystem(opts, filepath):

    try:
        api = Trading(debug=opts.debug, config_file=opts.yaml, appid=opts.appid, domain=opts.domain,
                      certid=opts.certid, devid=opts.devid, warnings=True)

        # pass in an open file
        # the Requests module will close the file
        files = {'file': ('EbayImage', open(filepath, 'rb'))}

        pictureData = {
            "WarningLevel": "High",
            "PictureName": "WorldLeaders"
        }

        api.execute('UploadSiteHostedPictures', pictureData, files=files)
        dump(api)

    except ConnectionError as e:
        print(e)
        print(e.response.dict())


def memberMessages(opts):

    try:
        api = Trading(debug=opts.debug, config_file=opts.yaml, appid=opts.appid, domain=opts.domain,
                      certid=opts.certid, devid=opts.devid, warnings=True)

        now = datetime.datetime.now()

        memberData = {
            "WarningLevel": "High",
            "MailMessageType": "All",
            # "MessageStatus": "Unanswered",
            "StartCreationTime": now - datetime.timedelta(days=60),
            "EndCreationTime": now,
            "Pagination": {
                "EntriesPerPage": "5",
                "PageNumber": "1"
            }
        }

        api.execute('GetMemberMessages', memberData)

        dump(api)

        if api.response.reply.has_key('MemberMessage'):
            messages = api.response.reply.MemberMessage.MemberMessageExchange

            if type(messages) != list:
                messages = [messages]

            for m in messages:
                print("%s: %s" % (m.CreationDate, m.Question.Subject[:50]))

    except ConnectionError as e:
        print(e)
        print(e.response.dict())


def getUser(opts):
    try:

        api = Trading(debug=opts.debug, config_file=opts.yaml, appid=opts.appid, domain=opts.domain,
                      certid=opts.certid, devid=opts.devid, warnings=True, timeout=20, siteid='101')

        api.execute('GetUser', {'UserID': 'sallyma789'})
        dump(api, full=False)

    except ConnectionError as e:
        print(e)
        print(e.response.dict())


def getOrders(opts):

    try:
        api = Trading(debug=opts.debug, config_file=opts.yaml, appid=opts.appid, domain=opts.domain,
                      certid=opts.certid, devid=opts.devid, warnings=True, timeout=20)

        api.execute('GetOrders', {'NumberOfDays': 30})
        dump(api, full=False)

    except ConnectionError as e:
        print(e)
        print(e.response.dict())


def categories(opts):

    try:
        api = Trading(debug=opts.debug, config_file=opts.yaml, appid=opts.appid, domain=opts.domain,
                      certid=opts.certid, devid=opts.devid, warnings=True, timeout=20, siteid='0')

        callData = {
            'DetailLevel': 'ReturnAll',
            'CategorySiteID': 101,
            'LevelLimit': 4,
        }

        api.execute('GetCategories', callData)
        dump(api, full=False)

    except ConnectionError as e:
        print(e)
        print(e.response.dict())

'''
api = trading(domain='api.sandbox.ebay.com')
api.execute('GetCategories', {
    'DetailLevel': 'ReturnAll',
    'CategorySiteID': 101,
    'LevelLimit': 4,
})
'''

if __name__ == "__main__":
    (opts, args) = init_options()

    print("Trading API Samples for version %s" % ebaysdk.get_version())

    run(opts)
    feedback(opts)
    verifyAddItem(opts)
    getTokenStatus(opts)
    verifyAddItemErrorCodes(opts)
    uploadPicture(opts)
    uploadPictureFromFilesystem(opts, ("%s/test_image.jpg" % os.path.dirname(__file__)))
    memberMessages(opts)
    categories(opts)

    # getUser(opts)
    # getOrders(opts)
